doc/guides/abi-stability.md

```diff
@@ -0,0 +1,107 @@
+# ABI Stability
+
+## Introduction
+An Application Binary Interface (ABI) is the compiled version of an Application
+Programming Interface (API). That is, the headers files describing the classes,
+functions, data structures, enumerations, and constants used to enable an
+application to perform a desired task correspond by way of compilation to a set
+of addresses and expected parameter values and memory structure sizes and
+layouts with which the provider of the ABI was compiled.
+
+The application using the ABI must be compiled such that the available
+addresses, expected parameter values, and memory structure sizes and layouts
+agree with those with which the ABI provider was compiled. This is usually
+accomplished by compiling against the headers provided by the ABI provider.
+
+Since the ABI provider and the ABI user may be compiled at different times with
```

Comment:
`ABI provider ('the library being required') and ABI user ('the consuming code')` maybe?

Reply:
I wouldn't say "the library being required", because in the case of a native addon, it's actually the executable program that is being required.

Reply:
I also wouldn't say "consuming" either, because that's too abstract. I can change it to "provider of the ABI" and "user of the ABI" to make it more down-to-Earth.